src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java

```diff
@@ -607,216 +510,72 @@ private BlockPos getNextBlockInShaftToMine()
         if (walkToBuilding())
         {
             return MINER_BUILDING_SHAFT;
         }
-        if (buildNextBlockInShaft())
-        {
-            return MINER_BUILDING_SHAFT;
-        }
-        getOwnBuilding().resetStartingLevelShaft();
 
-        return START_WORKING;
+        final BlockPos ladderPos = getOwnBuilding().getLadderLocation();
+        final int lastLadder = getLastLadder(ladderPos) + 1;
+
+        final int xOffset = SHAFT_RADIUS * getOwnBuilding().getVectorX();
+        final int zOffset = SHAFT_RADIUS * getOwnBuilding().getVectorZ();
+
+        initStructure(null, 0, new BlockPos(ladderPos.getX() + xOffset, lastLadder, ladderPos.getZ() + zOffset));
+
+        return CLEAR_STEP;
     }
 
     @NotNull
-    private AIState doNodeMining()
+    private AIState executeNodeMining()
     {
         @Nullable final Level currentLevel = getOwnBuilding().getCurrentLevel();
         if (currentLevel == null)
         {
             Log.getLogger().warn("Current Level not set, resetting...");
             getOwnBuilding().setCurrentLevel(getOwnBuilding().getNumberOfLevels() - 1);
-            return doNodeMining();
+            return executeNodeMining();
         }
-        mineAtLevel(currentLevel);
-        return MINER_CHECK_MINESHAFT;
+        return searchANodeToMine(currentLevel);
     }
 
-    private void mineAtLevel(@NotNull final Level currentLevel)
+    private AIState searchANodeToMine(@NotNull final Level currentLevel)
     {
-        if (workingNode == null)
+        if (workingNode == null || workingNode.getStatus() == Node.NodeStatus.COMPLETED)
         {
-            workingNode = findNodeOnLevel(currentLevel);
-            return;
+            workingNode = currentLevel.getRandomNode();
+            return MINER_CHECK_MINESHAFT;
         }
-        //Looking for a node to stand on while mining workingNode
-        int foundDirection = 0;
-        @Nullable Node foundNode = null;
-        @NotNull final List<Integer> directions = Arrays.asList(1, 2, 3, 4);
 
-        for (final Integer dir : directions)
-        {
-            final Optional<Node> node = tryFindNodeInDirectionOfNode(currentLevel, workingNode, dir);
-            if (node.isPresent() && getNodeStatusForDirection(node.get(), invertDirection(dir)) == Node.NodeStatus.COMPLETED)
-            {
-                foundDirection = dir;
-                foundNode = node.get();
-                break;
-            }
-        }
-        if (foundNode == null || foundDirection <= 0)
-        {
-            workingNode = null;
-            return;
-        }
-        int xOffSet = getXDistance(foundDirection) / 2;
-        int zOffSet = getZDistance(foundDirection) / 2;
-        if (xOffSet > 0)
-        {
-            xOffSet += 1;
-        }
-        else
+        //normal facing +x
+        int rotation = 0;
+
+        int vectorX = workingNode.getX() < workingNode.getParent().getX() ? -1 : (workingNode.getX() > workingNode.getParent().getX() ? 1 : 0);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'vectorX' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)